### PR TITLE
[Build] Support Samsung VD product build system

### PR DESCRIPTION
**Pull request by again4you**
This patch supports the Samsung VD product build system.
Below features are off since VD products do not support them.

* Privacy / Privilege Manager
* Resource Manager
* ProtoBuffer / FlatBuffer
* Disable Test since python-numpy is not supported

~~Moreover, NNFW is temporarily disabled since it is not ready yet.~~

Signed-off-by: Sangjung Woo <sangjung.woo@samsung.com>

**Self evaluation:**
1. Build test: [*]Passed [ ]Failed []Skipped
2. Run test: [*]Passed [ ]Failed []Skipped

**Related Issue**
* https://github.com/nnstreamer/nnstreamer/issues/2547

#### Submit v5
* Enable `nnfw`

#### Submit v4
* Cleanup spec file for VD build

#### Submit v3
* Add omitted `enable_tizen_feature` condition
* Clean up meson.build & meson_options.txt files

#### Submit v2
* Add `enable_tizen_privilege_check` and `enable_tizen_feature_check` options
* Cleanup source code and build script (i.e. meson.build)

#### Submit v1
* First Draft

### How to test
* Set your .gbs.conf as below.
```bash
[profile.tizen]
obs = obs.tizen
repos = repo.tizen_tv, repo.tizen_tv_base
buildroot = ~/GBS-ROOT/

[obs.tizen]
url = https://api.tizen.org

[repo.tizen_tv]
url=http://168.219.245.245/products/tv/official/2021/ONEMAIN/NikeM/latest/repos/product/armv7l/packages/

[repo.tizen_tv_base]
url=http://168.219.245.245/base/tizen-6.0-base/standard/latest/repos/base/armv7l/packages/
```

### CODE REVIEW

File: api/capi/include/nnstreamer-capi-private.h [hunk 1]
```diff
@@ -62,6 +64,14 @@ typedef enum { MM_RESOURCE_MANAGER_RES_TYPE_MAX } mm_resource_manager_res_type_e
 #error Tizen version is not defined.
 #endif
 
+#else
+/* In case of VD products, below functions are not supported. */
+#define convert_tizen_element(...) ML_ERROR_NONE
+#define get_tizen_resource(...) ML_ERROR_NONE
+#define release_tizen_resource(...)
```
Comment (myungjoo) [inline api/capi/include/nnstreamer-capi-private.h:71]: Enclosing them with do { } while(0) is safer.

File: meson_options.txt [hunk 2]
```diff
@@ -37,4 +37,5 @@ option('enable-tizen-sensor', type: 'boolean', value: false)
 option('enable-edgetpu', type: 'boolean', value: false)
 option('enable-openvino', type: 'boolean', value: false)
 option('enable-vivante', type: 'boolean', value: false)
+option('enable-vd-product', type: 'boolean', value: false)
```
Comment (myungjoo) [inline meson_options.txt:40]: let's name it ```enable-tizen-permission-check```, with default value: true.
For VD, use ```-Denable-tizen-permission-check=false```

File: meson.build [hunk 3]
```diff
@@ -31,6 +31,11 @@ if get_option('enable-tizen')
   tizenVmajor = get_option('tizen-version-major')
   add_project_arguments('-DTIZENVERSION='+tizenVmajor.to_string(), language: ['c', 'cpp'])
 
+  # In case of VD products, some packages are not supported.
+  if get_option('enable-vd-product')
+    add_project_arguments('-D__VD_PRODUCT_SUPPORT__', language: ['c', 'cpp'])
```
Comment (myungjoo) [inline meson.build:36]: Does this depend on __TIZEN__? or not?

File: packaging/nnstreamer.spec [hunk 4]
```diff
@@ -9,6 +9,15 @@
 %define		flatbuf_support 1
 %define		protobuf_support 1
 %define		nnfw_support 1
+%define		enable_test 1
+
+%if ("%{_vd_cfg_product_type}" == "AUDIO") || ("%{_vd_cfg_product_type}" == "TV") || ("%{_vd_cfg_product_type}" == "LFD") || ("%{_vd_cfg_product_type}" == "HTV") || ("%{_vd_cfg_product_type}" == "AV") || ("%{_vd_cfg_product_type}" == "IWB") || ("%{_vd_cfg_product_type}" == "WALL")
```
Comment (myungjoo) [inline packaging/nnstreamer.spec:14]: Or... simply..
```
%if "%{?profile}" == "tv"
```

File: packaging/nnstreamer.spec [hunk 5]
```diff
@@ -9,6 +9,15 @@
...
+%define		samsung_vd_product_support  1
+%define		enable_vd_product -Denable-vd-product=true
+%else
+%define		samsung_vd_product_support  0
+%define		enable_vd_product -Denable-vd-product=false
+%endif
```
Comment (myungjoo) [inline packaging/nnstreamer.spec:20]: Rename variables... 
samsung_vd_product_support --> disable_extra_subplugins
enable_vd_product --> disable_tizen_permission_check

File: packaging/nnstreamer.spec [hunk 6]
```diff
@@ -496,8 +528,8 @@ meson --buildtype=plain --prefix=%{_prefix} --sysconfdir=%{_sysconfdir} --libdir
 	--bindir=%{nnstexampledir} --includedir=%{_includedir} -Dinstall-example=true \
 	%{enable_api} %{enable_tizen} %{element_restriction} -Denable-env-var=false -Denable-symbolic-link=false \
 	%{enable_tf_lite} %{enable_tf} %{enable_pytorch} %{enable_caffe2} %{enable_python} \
-	%{enable_nnfw_runtime} %{enable_mvncsdk2} %{enable_armnn} %{enable_edgetpu}  %{enable_vivante} %{enable_flatbuf} \
-	%{enable_tizen_sensor} %{enable_test_coverage} \
+	%{enable_nnfw_runtime} %{enable_mvncsdk2} %{enable_armnn} %{enable_edgetpu}  %{enable_vivante} %{enable_flatbuf} ${enable_vd_product} \
+	%{enable_tizen_sensor} %{enable_test} %{enable_test_coverage} \
```
Comment (myungjoo) [inline packaging/nnstreamer.spec:532]: enable_test is 0 or 1. Why is it here?

File: api/capi/src/nnstreamer-capi-tizen.c [hunk 7]
```diff
@@ -21,7 +21,7 @@
  * @bug No known bugs except for NYI items
  */
 
-#if !defined (__TIZEN__)
+#if !defined (__TIZEN__) && !defined (__VD_PRODUCT_SUPPORT__)
```
Comment (myungjoo) [inline api/capi/src/nnstreamer-capi-tizen.c:24]: Not necessary.

File: api/capi/src/nnstreamer-capi-tizen-feature.c [hunk 8]
```diff
@@ -0,0 +1,142 @@
...
+ * version 2.1 of the License.
+ *
+ * This library is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the GNU
+ * Library General Public License for more details.
+ */
+
+/**
+ * @file nnstreamer-capi-tizen-feature.c
```
Comment (myungjoo) [inline api/capi/src/nnstreamer-capi-tizen-feature.c:16]: filename rename:
nnstreamer-capi-tizen-permission-check.c ?

File: packaging/nnstreamer.spec (hunk 4 above)
Comment (again4you) [reply packaging/nnstreamer.spec:14]: We can't. Public Tizen TV profile is different from VD product.
I don't know it is the right approach but other packages use this way.
I will check it later.

File: packaging/nnstreamer.spec (hunk 6 above)
Comment (again4you) [reply packaging/nnstreamer.spec:532]: IIRC, some test cases use Python and `python-numpy` is not supported. I will check it.

File: api/capi/src/nnstreamer-capi-tizen.c (hunk 7 above)
Comment (again4you) [reply api/capi/src/nnstreamer-capi-tizen.c:24]: ACR!

File: api/capi/src/nnstreamer-capi-tizen-feature.c (hunk 8 above)
Comment (again4you) [reply api/capi/src/nnstreamer-capi-tizen-feature.c:16]: ACK!

File: meson_options.txt (hunk 2 above)
Comment (again4you) [reply meson_options.txt:40]: Then, I will add two options.
* enable-tizen-permission-check
* enable-tizen-feature-check

File: packaging/nnstreamer.spec (hunk 5 above)
Comment (again4you) [reply packaging/nnstreamer.spec:20]: ACK!

File: api/capi/meson.build [hunk 9]
```diff
@@ -32,8 +32,12 @@ capi_main += join_paths(meson.current_source_dir(), 'src', 'nnstreamer-capi-util
...
 capi_main += join_paths(meson.current_source_dir(), 'src', 'tensor_filter_single.c')
 
-if get_option('enable-tizen')
-  capi_main += join_paths('src', 'nnstreamer-capi-tizen.c')
+if get_option('enable-tizen') and get_option('enable-tizen-feature-check')
+  capi_main += join_paths(meson.current_source_dir(), 'src', 'nnstreamer-capi-tizen-feature-check.c')
+endif
+
+if get_option('enable-tizen') and get_option('enable-tizen-privilege-check')
+  capi_main += join_paths('src', 'nnstreamer-capi-tizen-privilege-check.c')
```
Comment (myungjoo) [inline api/capi/meson.build:40]: ```meson.current_source_dir()``` is missing?

File: meson_options.txt [hunk 10]
```diff
@@ -37,4 +37,7 @@ option('enable-tizen-sensor', type: 'boolean', value: false)
 option('enable-edgetpu', type: 'boolean', value: false)
 option('enable-openvino', type: 'boolean', value: false)
 option('enable-vivante', type: 'boolean', value: false)
+option('enable-vivante', type: 'boolean', value: false)
```
Comment (myungjoo) [inline meson_options.txt:40]: duplicated line?

File: packaging/nnstreamer.spec [hunk 11]
```diff
@@ -386,6 +411,9 @@ You may enable this package to use Google Edge TPU with NNStreamer and Tizen ML
 %define enable_mvncsdk2 -Dmvncsdk2-support=disabled
 %define enable_nnfw_runtime -Dnnfw-runtime-support=disabled
 %define element_restriction -Denable-element-restriction=false
+%define enable_tizen_privilege_check -Denable-tizen-privilege-check=true
+%define enable_tizen_feature_check -Denable-tizen-feature-check=true
```
Comment (myungjoo) [inline packaging/nnstreamer.spec:415]: A statement to control ```enable_tizen_feature_check ``` is missing.

**Comment by again4you** [comment]
To detect VD product build, Semun Lee guides me that the below code is used in other packages.
```bash
%if ("%{_vd_cfg_product_type}" == "AUDIO") || ("%{_vd_cfg_product_type}" == "TV") || ("%{_vd_cfg_product_type}" == "LFD") || ("%{_vd_cfg_product_type}" == "HTV") || ("%{_vd_cfg_product_type}" == "AV") || ("%{_vd_cfg_product_type}" == "IWB") || ("%{_vd_cfg_product_type}" == "WALL")
  # Enable VD Product Build Option
  ...
%endif
```

However, @myungjoo  advices me that simple `%if "%{?profile}" == "tv"` is enough for this work.
After setting gbs configuration for VD, I check that it builds as I intended.

Could you please check what is the best approach for this work?

**Comment by semun-lee** [comment]
> To detect VD product build, Semun Lee guides me that the below code is used in other packages.
> 
> ```shell
> %if ("%{_vd_cfg_product_type}" == "AUDIO") || ("%{_vd_cfg_product_type}" == "TV") || ("%{_vd_cfg_product_type}" == "LFD") || ("%{_vd_cfg_product_type}" == "HTV") || ("%{_vd_cfg_product_type}" == "AV") || ("%{_vd_cfg_product_type}" == "IWB") || ("%{_vd_cfg_product_type}" == "WALL")
>   # Enable VD Product Build Option
>   ...
> %endif
> ```
> 
> However, @myungjoo advices me that simple `%if "%{?profile}" == "tv"` is enough for this work.
> After setting gbs configuration for VD, I check that it builds as I intended.
> 
> Could you please check what is the best approach for this work?

It might be enough because we don't use profile macro in public unified build.

File: packaging/nnstreamer.spec (hunk 11 above)
Comment (again4you) [reply packaging/nnstreamer.spec:415]: ACK!

File: meson_options.txt (hunk 10 above)
Comment (again4you) [reply meson_options.txt:40]: ACK!

File: api/capi/meson.build (hunk 9 above)
Comment (again4you) [reply api/capi/meson.build:40]: ACK!

**Comment by myungjoo** [comment]
> > To detect VD product build, Semun Lee guides me that the below code is used in other packages.
> > ```shell
> > %if ("%{_vd_cfg_product_type}" == "AUDIO") || ("%{_vd_cfg_product_type}" == "TV") || ("%{_vd_cfg_product_type}" == "LFD") || ("%{_vd_cfg_product_type}" == "HTV") || ("%{_vd_cfg_product_type}" == "AV") || ("%{_vd_cfg_product_type}" == "IWB") || ("%{_vd_cfg_product_type}" == "WALL")
> >   # Enable VD Product Build Option
> >   ...
> > %endif
> > ```
> > 
> > 
> > However, @myungjoo advices me that simple `%if "%{?profile}" == "tv"` is enough for this work.
> > After setting gbs configuration for VD, I check that it builds as I intended.
> > Could you please check what is the best approach for this work?
> 
> It might be enough because we don't use profile macro in public unified build.

Actually, ```%{?profile}``` was guided by PM-TV.
And we can use ```profile``` because Tizen:Unified does NOT use it 👍

File: packaging/nnstreamer.spec [hunk 12]
```diff
@@ -9,6 +9,14 @@
 %define		flatbuf_support 1
 %define		protobuf_support 1
 %define		nnfw_support 1
+%define		check_test 1
+%define		enable_tizen_privilege 1
+%define		enable_tizen_feature 1
+%define		enable_extra_subplugins 1
+
+%if "%{?profile}" == "tv"
+%define		enable_extra_subplugins 0
```
Comment (myungjoo) [inline packaging/nnstreamer.spec:18]:
```suggestion
%define		enable_extra_subplugins 0
%define		enable_tizen_privilege 0
%define		enable_tizen_feature 0
```

File: packaging/nnstreamer.spec [hunk 13]
```diff
@@ -50,6 +58,18 @@
...
+%if !0%{?enable_extra_subplugins}
+%define		flatbuf_support 0
+%define		protobuf_support 0
+%define		python_support 0
+%define		mvncsdk2_support 0
+%define		edgetpu_support 0
+%define		check_test 0
+%define		enable_tizen_privilege 0
+# temporarily disable NNFW since it is not ready
+%define		nnfw_support 0
```
Comment (myungjoo) [inline packaging/nnstreamer.spec:70]:
```
%define		check_test 0
%define		enable_tizen_privilege 0
```
These are not subplugins.

File: packaging/nnstreamer.spec (hunk 12 above)
Comment (again4you) [reply packaging/nnstreamer.spec:18]: I understood what you say. I will.

File: packaging/nnstreamer.spec (hunk 13 above)
Comment (again4you) [reply packaging/nnstreamer.spec:70]: ACK!

**Comment by again4you** [comment]
Because of `nnstreamer_protobuf`, below Ubuntu buildbreak occurs.
https://github.com/nnstreamer/nnstreamer/pull/2571 PR fixes this bug.

```bash
...
In file included from /var/www/html/nnstreamer/TAOS-CI/ci/repo-workers/pr-checker/2554-202007241111210.013263940811157-fa83286b2c9a28959b8ed611bf469e8831c477d7/nnstreamer/ext/nnstreamer/tensor_decoder/tensordec-protobuf.cc:17:0:
../ext/nnstreamer/extra/nnstreamer_protobuf.h:37:1: error: 'GstFlowReturn' does not name a type
 GstFlowReturn gst_tensor_decoder_protobuf (const GstTensorsConfig *config,
 ^
../ext/nnstreamer/extra/nnstreamer_protobuf.h:49:1: error: 'GstBuffer' does not name a type
 GstBuffer *gst_tensor_converter_protobuf (GstBuffer *in_buf, gsize *frame_size,
...
```

**Pull request merged**